File: packages/haptics/ios/Plugin/Classes/Options/PlayPatternOptions.swift
import Foundation
import Capacitor

@objc public class PlayPatternOptions: NSObject {
    struct Event {
        let duration: Double?
        let intensity: Float
        let sharpness: Float
        let time: Double
    }

    let events: [Event]

    init(_ call: CAPPluginCall) throws {
        self.events = try PlayPatternOptions.getEventsFromCall(call)
    }

    private static func getEventsFromCall(_ call: CAPPluginCall) throws -> [Event] {
        guard let eventObjects = call.getArray("events", JSObject.self), !eventObjects.isEmpty else {
            throw CustomError.eventsMissing
        }
        var events: [Event] = []
        for eventObject in eventObjects {
            guard let time = (eventObject["time"] as? NSNumber)?.doubleValue else {
                throw CustomError.timeMissing
            }
            guard time >= 0 else {
                throw CustomError.timeInvalid
            }
            guard let intensity = (eventObject["intensity"] as? NSNumber)?.floatValue else {
                throw CustomError.intensityMissing
            }
            guard intensity >= 0, intensity <= 1 else {
                throw CustomError.intensityInvalid
            }
            var sharpness: Float = 0.5
            if let value = (eventObject["sharpness"] as? NSNumber)?.floatValue {
                guard value >= 0, value <= 1 else {
                    throw CustomError.sharpnessInvalid
                }
                sharpness = value
            }
            var duration: Double?
            if let value = (eventObject["duration"] as? NSNumber)?.doubleValue {
                guard value > 0 else {
                    throw CustomError.durationInvalid
                }
                duration = value
            }
            events.append(Event(duration: duration, intensity: intensity, sharpness: sharpness, time: time))
        }
        return events
    }
}

File: packages/haptics/android/src/main/java/io/capawesome/capacitorjs/plugins/haptics/classes/options/PlayPatternOptions.java
package io.capawesome.capacitorjs.plugins.haptics.classes.options;

import androidx.annotation.NonNull;
import com.getcapacitor.JSArray;
import com.getcapacitor.PluginCall;
import io.capawesome.capacitorjs.plugins.haptics.classes.CustomExceptions;
import java.util.ArrayList;
import java.util.List;
import org.json.JSONObject;

public class PlayPatternOptions {

    public static class Event {

        private final long durationMs;
        private final float intensity;
        private final long timeMs;

        public Event(long durationMs, float intensity, long timeMs) {
            this.durationMs = durationMs;
            this.intensity = intensity;
            this.timeMs = timeMs;
        }

        public long getDurationMs() {
            return durationMs;
        }

        public float getIntensity() {
            return intensity;
        }

        public long getTimeMs() {
            return timeMs;
        }

        public boolean isTransient() {
            return durationMs == 0;
        }
    }

    @NonNull
    private final List<Event> events;

    public PlayPatternOptions(@NonNull PluginCall call) throws Exception {
        this.events = PlayPatternOptions.getEventsFromCall(call);
    }

    @NonNull
    public List<Event> getEvents() {
        return events;
    }

    @NonNull
    private static List<Event> getEventsFromCall(@NonNull PluginCall call) throws Exception {
        JSArray eventsArray = call.getArray("events");
        if (eventsArray == null || eventsArray.length() == 0) {
            throw CustomExceptions.EVENTS_MISSING;
        }
        List<Event> events = new ArrayList<>();
        for (int index = 0; index < eventsArray.length(); index++) {
            JSONObject eventObject = eventsArray.getJSONObject(index);
            if (!eventObject.has("time")) {
                throw CustomExceptions.TIME_MISSING;
            }
            double time = eventObject.getDouble("time");
            if (time < 0) {
                throw CustomExceptions.TIME_INVALID;
            }
            if (!eventObject.has("intensity")) {
                throw CustomExceptions.INTENSITY_MISSING;
            }
            double intensity = eventObject.getDouble("intensity");
            if (intensity < 0 || intensity > 1) {
                throw CustomExceptions.INTENSITY_INVALID;
            }
            long durationMs = 0;
            if (eventObject.has("duration")) {
                double duration = eventObject.getDouble("duration");
                if (duration <= 0) {
                    throw CustomExceptions.DURATION_INVALID;
                }
                durationMs = Math.round(duration * 1000);
            }
            events.add(new Event(durationMs, (float) intensity, Math.round(time * 1000)));
        }
        return events;
    }
}
